File: mm-cs/src/MetaMessage/Core/TagFieldParser.cs
using ValueType = MetaMessage.Ir.ValueType;
using MetaMessage.Ir;

namespace MetaMessage.Core;

public static class TagFieldParser
{
    public static Tag Parse(byte[] data)
    {
        Tag tag = Tag.Empty();
        int offset = 0;

        while (offset < data.Length)
        {
            if (offset >= data.Length)
                break;

            byte header = data[offset++];

            var key = (TagKey)(header & 0xF8);
            int lenInfo = header & 0x07;

            switch (key)
            {
                case TagKey.KExample:
                    tag.Example = true;
                    break;

                case TagKey.KIsNull:
                    tag.IsNull = true;
                    tag.Nullable = true;
                    break;

                case TagKey.KNullable:
                    tag.Nullable = true;
                    break;

                case TagKey.KDeprecated:
                    tag.Deprecated = true;
                    break;

                case TagKey.KAllowEmpty:
                    tag.AllowEmpty = true;
                    break;

                case TagKey.KUnique:
                    tag.Unique = true;
                    break;

                case TagKey.KChildNullable:
                    tag.ChildNullable = true;
                    break;

                case TagKey.KChildAllowEmpty:
                    tag.ChildAllowEmpty = true;
                    break;

                case TagKey.KChildUnique:
                    tag.ChildUnique = true;
                    break;

                case TagKey.KDesc:
                    {
                        int strLen = DecodeStringLength(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag desc data overflow");
                        tag.Desc = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KType:
                    {
                        if (offset >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.Type = (ValueType)data[offset++];
                        break;
                    }

                case TagKey.KDefault:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag default value data overflow");
                        tag.DefaultVal = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KMin:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag min data overflow");
                        tag.Min = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KMax:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag max data overflow");
                        tag.Max = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KSize:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.Size = (int)DecodeU64(data, ref offset, byteCount);
                        break;
                    }

                case TagKey.KEnum:
                    {
                        int strLen = DecodeStringLength(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag enum data overflow");
                        tag.Enums = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        tag.Type = ValueType.Enums;
                        offset += strLen;
                        break;
                    }

                case TagKey.KPattern:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag pattern data overflow");
                        tag.Pattern = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KLocation:
                    {
                        int strLen = lenInfo;
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag location data overflow");
                        tag.Location = int.Parse(System.Text.Encoding.UTF8.GetString(data, offset, strLen));
                        offset += strLen;
                        break;
                    }

                case TagKey.KVersion:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.Version = (int)DecodeU64(data, ref offset, byteCount);
                        break;
                    }

                case TagKey.KMime:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.Mime = DecodeU64(data, ref offset, byteCount).ToString();
                        tag.Type = ValueType.Media;
                        break;
                    }

                case TagKey.KChildDesc:
                    {
                        int strLen = DecodeStringLength(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child desc data overflow");
                        tag.ChildDesc = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildType:
                    {
                        if (offset >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.ChildType = (ValueType)data[offset++];
                        break;
                    }

                case TagKey.KChildDefaultVal:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child default value data overflow");
                        tag.ChildDefaultVal = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildMin:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child min data overflow");
                        tag.ChildMin = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildMax:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child max data overflow");
                        tag.ChildMax = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildSize:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.ChildSize = (int)DecodeU64(data, ref offset, byteCount);
                        break;
                    }

                case TagKey.KChildEnums:
                    {
                        int strLen = DecodeStringLength(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child enum data overflow");
                        tag.ChildEnums = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildPattern:
                    {
                        int strLen = DecodeStringLengthSimple(data, ref offset, lenInfo);
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child pattern data overflow");
                        tag.ChildPattern = System.Text.Encoding.UTF8.GetString(data, offset, strLen);
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildLocation:
                    {
                        int strLen = lenInfo;
                        if (offset + strLen > data.Length)
                            throw new MmDecodeException("Tag child location data overflow");
                        tag.ChildLocation = int.Parse(System.Text.Encoding.UTF8.GetString(data, offset, strLen));
                        offset += strLen;
                        break;
                    }

                case TagKey.KChildVersion:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.ChildVersion = (int)DecodeU64(data, ref offset, byteCount);
                        break;
                    }

                case TagKey.KChildMime:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.ChildMime = DecodeU64(data, ref offset, byteCount).ToString();
                        tag.ChildType = ValueType.Media;
                        break;
                    }

                case TagKey.KMore:
                    {
                        int byteCount = lenInfo + 1;
                        if (offset + byteCount > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        tag.More = (int)DecodeU64(data, ref offset, byteCount);
                        break;
                    }

                default:
                    if (lenInfo <= 5)
                    {
                        if (offset + lenInfo > data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        offset += lenInfo;
                    }
                    else if (lenInfo == 6)
                    {
                        if (offset >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        int skipLen = data[offset++];
                        if (offset + skipLen > data.Length)
                            throw new MmDecodeException("Tag data overflow");
                        offset += skipLen;
                    }
                    else if (lenInfo == 7)
                    {
                        if (offset + 1 >= data.Length)
                            throw new MmDecodeException("Unexpected end of tag data");
                        int skipLen = (data[offset] << 8) | data[offset + 1];
                        offset += 2;
                        if (offset + skipLen > data.Length)
                            throw new MmDecodeException("Tag data overflow");
                        offset += skipLen;
                    }
                    break;
            }
        }

        return tag;
    }

    private static int DecodeStringLength(byte[] data, ref int offset, int lenInfo)
    {
        if (lenInfo <= 5)
        {
            return lenInfo;
        }

        if (lenInfo == 6)
        {
            if (offset >= data.Length)
                throw new MmDecodeException("Unexpected end of tag data");
            int len = data[offset];
            offset++;
            return len;
        }

        if (offset + 1 >= data.Length)
            throw new MmDecodeException("Unexpected end of tag data");
        int lenHi = data[offset];
        offset++;
        int lenLo = data[offset];
        offset++;
        return (lenHi << 8) | lenLo;
    }

    private static int DecodeStringLengthSimple(byte[] data, ref int offset, int lenInfo)
    {
        if (lenInfo < 7)
        {
            return lenInfo;
        }

        if (offset >= data.Length)
            throw new MmDecodeException("Unexpected end of tag data");
        int len = data[offset];
        offset++;
        return len;
    }

    private static ulong DecodeU64(byte[] data, ref int offset, int byteCount)
    {
        ulong result = 0;
        for (int i = 0; i < byteCount; i++)
        {
            if (offset >= data.Length)
                throw new MmDecodeException("Unexpected end of tag data");
            result = (result << 8) | data[offset];
            offset++;
        }
        return result;
    }
}

File: mm-kt/src/main/kotlin/io/github/metamessage/core/TagFieldParser.kt
package io.github.metamessage.core

import io.github.metamessage.ir.Mime
import io.github.metamessage.ir.Tag
import io.github.metamessage.ir.ValueType
import java.nio.charset.StandardCharsets

object TagFieldParser {
    fun parseOne(c: Cursor, tag: Tag): Int {
        val start = c.pos
        val b = c.read()
        val prefix = b and 0xF8
        val low = b and 0x07
        when (prefix) {
            Tag.TagKey.K_IS_NULL -> {
                tag.isNull = (low and 1) == 1
                if (tag.isNull) tag.nullable = true
            }
            Tag.TagKey.K_EXAMPLE -> tag.example = (low and 1) == 1
            Tag.TagKey.K_DESC -> readSizedUtf8(c, tag, low, 0)
            Tag.TagKey.K_TYPE -> tag.type = ValueType.fromCode(c.read())
            Tag.TagKey.K_DEPRECATED -> tag.deprecated = (low and 1) == 1
            Tag.TagKey.K_NULLABLE -> tag.nullable = (low and 1) == 1
            Tag.TagKey.K_ALLOW_EMPTY -> tag.allowEmpty = (low and 1) == 1
            Tag.TagKey.K_UNIQUE -> tag.unique = (low and 1) == 1
            Tag.TagKey.K_DEFAULT_VAL -> tag.default_val = readShortUtf8(c, low)
            Tag.TagKey.K_MIN -> tag.min = readShortUtf8(c, low)
            Tag.TagKey.K_MAX -> tag.max = readShortUtf8(c, low)
            Tag.TagKey.K_SIZE -> tag.size = readUintN(c, low)
            Tag.TagKey.K_ENUM -> {
                tag.type = ValueType.ENUMS
                tag.enums = readSizedUtf8Only(c, low)
            }
            Tag.TagKey.K_PATTERN -> tag.pattern = readShortUtf8(c, low)
            Tag.TagKey.K_LOCATION -> tag.location = readAscii(c, low).toInt()
            Tag.TagKey.K_VERSION -> tag.version = readUintN(c, low)
            Tag.TagKey.K_MIME -> readMime(c, tag, low, true)
            Tag.TagKey.K_CHILD_DESC -> readSizedUtf8(c, tag, low, 1)
            Tag.TagKey.K_CHILD_TYPE -> tag.childType = ValueType.fromCode(c.read())
            Tag.TagKey.K_CHILD_NULLABLE -> tag.childNullable = (low and 1) == 1
            Tag.TagKey.K_CHILD_ALLOW_EMPTY -> tag.childAllowEmpty = (low and 1) == 1
            Tag.TagKey.K_CHILD_UNIQUE -> tag.childUnique = (low and 1) == 1
            Tag.TagKey.K_CHILD_DEFAULT_VAL -> tag.childDefaultVal = readShortUtf8(c, low)
            Tag.TagKey.K_CHILD_MIN -> tag.childMin = readShortUtf8(c, low)
            Tag.TagKey.K_CHILD_MAX -> tag.childMax = readShortUtf8(c, low)
            Tag.TagKey.K_CHILD_SIZE -> tag.childSize = readUintN(c, low)
            Tag.TagKey.K_CHILD_ENUMS -> {
                tag.childType = ValueType.ENUMS
                tag.childEnums = readSizedUtf8Only(c, low)
            }
            Tag.TagKey.K_CHILD_PATTERN -> tag.childPattern = readShortUtf8(c, low)
            Tag.TagKey.K_CHILD_LOCATION -> tag.childLocation = readAscii(c, low).toInt()
            Tag.TagKey.K_CHILD_VERSION -> tag.childVersion = readUintN(c, low)
            Tag.TagKey.K_CHILD_MIME -> readMime(c, tag, low, false)
            else ->
                    throw MmDecodeException(
                            "invalid tag field prefix: 0x${Integer.toHexString(prefix)}"
                    )
        }
        return c.pos - start
    }

    private fun readSizedUtf8(c: Cursor, tag: Tag, low: Int, mode: Int) {
        val s = readSizedUtf8Only(c, low)
        if (mode == 0) tag.desc = s else tag.childDesc = s
    }

    private fun readSizedUtf8Only(c: Cursor, low: Int): String {
        return if (low <= 5) {
            String(c.readBytes(low), StandardCharsets.UTF_8)
        } else if (low == 6) {
            val l = c.read()
            String(c.readBytes(l), StandardCharsets.UTF_8)
        } else {
            val hi = c.read()
            val lo = c.read()
            val l = (hi shl 8) or lo
            String(c.readBytes(l), StandardCharsets.UTF_8)
        }
    }

    private fun readShortUtf8(c: Cursor, low: Int): String {
        return if (low < 7) {
            String(c.readBytes(low), StandardCharsets.UTF_8)
        } else {
            val l = c.read()
            String(c.readBytes(l), StandardCharsets.UTF_8)
        }
    }

    private fun readAscii(c: Cursor, low: Int): String {
        return String(c.readBytes(low), StandardCharsets.US_ASCII)
    }

    private fun readUintN(c: Cursor, low: Int): Int {
        require(low < 8) { "uint field length" }
        var v = 0
        for (i in 0..low) {
            v = (v shl 8) or c.read()
        }
        return v
    }

    private fun readMime(c: Cursor, tag: Tag, low: Int, self: Boolean) {
        val v = readUintN(c, low)
        if (self) {
            tag.type = ValueType.MEDIA
            tag.mime = Mime.toString(v)
        } else {
            tag.childType = ValueType.MEDIA
            tag.childMime = Mime.toString(v)
        }
    }

    class Cursor(var d: ByteArray, var pos: Int) {
        fun read(): Int {
            if (pos >= d.size) throw MmDecodeException("eof")
            return d[pos++].toInt() and 0xFF
        }

        fun readBytes(n: Int): ByteArray {
            if (pos + n > d.size) throw MmDecodeException("eof")
            val r = ByteArray(n)
            System.arraycopy(d, pos, r, 0, n)
            pos += n
            return r
        }
    }
}
